rust/relay/ebpf-turn-router/src/move_payload.rs

```diff
@@ -0,0 +1,306 @@
+    shift_forward_and_shrink::<{ Ipv4Hdr::LEN }>(ctx)
+}
+
+#[inline(always)]
+pub fn remove_channel_data_header_ipv6(ctx: &XdpContext) -> Result<(), Error> {
+    shift_forward_and_shrink::<{ Ipv6Hdr::LEN }>(ctx)
+}
+
+/// Extend tail by channel data header length and write channel data header at the front of the
+/// payload.
+#[inline(never)]
+fn extend_and_shift_backward<const IP_HEADER_LEN: usize>(
+    ctx: &XdpContext,
+    header: &CdHdr,
+) -> Result<(), Error> {
+    let off = payload_offset::<IP_HEADER_LEN>();
+
+    // Bounds check for verifier
+    if ctx.data() + off > ctx.data_end() {
+        return Err(Error::PacketTooShort);
+    }
+
+    // Extend tail
+    let ret = unsafe { bpf_xdp_adjust_tail(ctx.ctx, CdHdr::LEN as i32) };
+    if ret < 0 {
+        return Err(Error::XdpAdjustTailFailed(ret));
+    }
+
+    let data_size = ctx.data_end() - ctx.data();
+
+    // Copy payload
+    let original_data_size = data_size - CdHdr::LEN;
+    let payload_size = original_data_size - off;
+    let src_off = original_data_size;
+    let dst_off = data_size;
+
+    copy_payload_descending(ctx, src_off, dst_off, payload_size)?;
+
+    // Write header
+    let data_start = ctx.data();
+    let data_end = ctx.data_end();
+
+    // Verify the header write location is within bounds
+    if data_start + off + CdHdr::LEN > data_end {
+        return Err(Error::PacketTooShort);
+    }
+
+    let src_ptr = header as *const CdHdr;
+    let dst_ptr = data_start + off;
+    unsafe {
+        core::ptr::copy_nonoverlapping(src_ptr as *const u8, dst_ptr as *mut u8, CdHdr::LEN);
+    }
+
+    Ok(())
+}
+
+/// Slide payload toward head by `delta` and shrink tail by `delta`.
+#[inline(never)]
+fn shift_forward_and_shrink<const IP_HEADER_LEN: usize>(ctx: &XdpContext) -> Result<(), Error> {
+    let off = payload_offset::<IP_HEADER_LEN>();
+
+    if ctx.data() + off > ctx.data_end() {
+        return Err(Error::PacketTooShort);
+    }
+
+    let data_size = ctx.data_end() - ctx.data();
+
+    // Copy payload
+    let remaining = data_size - off - CdHdr::LEN;
+    let src_off = off + CdHdr::LEN;
+    let dst_off = off;
+
+    copy_payload_ascending(ctx, src_off, dst_off, remaining)?;
+
+    // Trim the now-unused bytes from the tail
+    let ret = unsafe { bpf_xdp_adjust_tail(ctx.ctx, -(CdHdr::LEN as i32)) };
+    if ret < 0 {
+        return Err(Error::XdpAdjustTailFailed(ret));
+    }
+
+    Ok(())
+}
+
+#[inline(always)]
+fn payload_offset<const IP_HEADER_LEN: usize>() -> usize {
+    EthHdr::LEN + IP_HEADER_LEN + UdpHdr::LEN
+}
+
+/// Shift payload forward (headwards) by `n` bytes
+///
+/// This is performed in a single pass using binary decomposition from 1024 down to 1 byte.
+/// This done because the verifier doesn't like non compile-time guarantees that pointers
+/// are within bounds, and the variable length payload can cause an issue.
+#[inline(never)]
+fn copy_payload_ascending(
```

Comment:
I think these two copy functions can be unified (if the verifier agrees):

- Make `copy_n` take `&mut src_off` and `&mut dst_off`
- If `src_off > dst_off`, copy forwards,
- else copy backwards
- `copy_n` can then modify both variables by `N`